pipeline_dp/dp_engine.py

```diff
@@ -288,6 +293,71 @@ def _add_empty_public_partitions(self, col, public_partitions,
             col, empty_accumulators,
             "Join public partitions with partitions from data")
 
+    def _bound_per_user_contributions(self, col, max_contributions: int,
+                                      aggregator_fn):
+        """Bounds the total contributions by a privacy_id.
+
+        Args:
+          col: collection, with types of each element: (privacy_id,
+            partition_key, value).
+          max_contributions: maximum number of records that one privacy id can
+            contribute.
+          aggregator_fn: function that takes a list of values and returns an
+            aggregator object which handles all aggregation logic.
+
+        return: collection with elements ((privacy_id, partition_key),
+              accumulator).
+        """
+        # bounds the total number of contributions by a privacy id.
+        col = self._backend.map_tuple(
+            col, lambda pid, pk, v: (pid, (pk, v)),
+            "Rekey to ((privacy_id), (partition_key, value))")
+        col = self._backend.sample_fixed_per_key(col, max_contributions,
+                                                 "Sample per privacy_id")
+        self._add_report_stage(
+            f"User contributions bounding: randomly selected not "
+            f"more than {max_contributions} contributions")
+
+        # (privacy_id, [(partition_key, value)])
+
+        # Convert the per privacy id list into a dict with key as partition_key
+        # and value to be list of input values.
+        def collect_values_per_partition_key_per_privacy_id(input_list):
+            d = collections.defaultdict(list)
+            for key, value in input_list:
+                d[key].append(value)
+            return d
+
+        col = self._backend.map_values(
+            col, collect_values_per_partition_key_per_privacy_id,
+            "Apply fn to convert per privacy id values into a dict of values"
```

Comment:
Nit: "Group per (privacy_id, partition_key)." 

It's better to keep stage name shorter, it's more convenient to check them in moninoring UI

Reply:
Done.